types/react/index.d.ts

```diff
@@ -77,34 +77,49 @@ declare namespace React {
         ref?: LegacyRef<T>;
     }
 
-    interface ReactElement<P> {
+    interface ReactElement<P, R = any> {
         type: string | ComponentClass<P> | FunctionComponent<P>;
         props: P;
         key: Key | null;
+        ref: R | null;
+    }
+
+    type RefTypeOfProps<P> = 'ref' extends keyof P
+        ? P extends { ref?: infer R }
```

Comment:
This is declared as part of the type of the DOM props (all of the intrinsic elements have their own `ref` prop); ref forwarding components are also implemented by adding a `ref` prop to the exposed props.

This is less than ideal because `key` and `ref` are never visible to the components themselves, but there's not really any better possible solution right now.